### PR TITLE
Fix dictwatcherdel crash

### LINKED ISSUE #11188: Crash with using dictwatcherdelete in dictwatcher callback
This causes a segfault:

```vim
let s:d = {}

function! W(...) abort
  call dictwatcherdel(s:d, '*', function('W'))
  call dictwatcherdel({}, 'meh', function('tr'))
endfunction

call dictwatcheradd(s:d, '*', function('W'))
let s:d.foo = 23
```

It's because the queue entry gets removed while using `QUEUE_FOREACH`, which is documented as undefined behavior: https://github.com/neovim/neovim/blob/3237901a472b459d89d06038cbfd78b997e95b96/src/nvim/lib/queue.h#L36-L40

Any suggestion how to address this?
Should `QUEUE_FOREACH` be fixed to handle this?  (I assume it might affect other areas also)

### PR DESCRIPTION
**Pull request by blueyed**
Fixes https://github.com/neovim/neovim/issues/11188.

### CODE REVIEW

**Comment by blueyed** [comment]
Hmm, appears to still crash on OpenBSD:
```
[ RUN      ] VimL dictionary notifications does not crash when using dictwatcherdel in callback: -- Tests exited non-zero: SIGPIPE
-- No output to stderr.
CMake Error at /home/build/neovim/cmake/RunTests.cmake:85 (message):
  functional tests failed with error: SIGPIPE
```

File: src/nvim/eval/typval.c [hunk 1]
```diff
@@ -1139,7 +1140,11 @@ bool tv_dict_watcher_remove(dict_T *const dict, const char *const key_pattern,
   }
 
   QUEUE_REMOVE(w);
```
Comment (blueyed) [inline src/nvim/eval/typval.c]: Does `QUEUE_REMOVE` need to be delayed also?

File: src/nvim/eval/typval.c [hunk 2]
```diff
@@ -1208,6 +1213,9 @@ void tv_dict_watcher_notify(dict_T *const dict, const char *const key,
       watcher->busy = true;
       callback_call(&watcher->callback, 3, argv, &rettv);
       watcher->busy = false;
+      if (watcher->needs_free) {
+        tv_dict_watcher_free(watcher);
+      }
```
Comment (justinmk) [inline src/nvim/eval/typval.c]: Does this mean `dictwatcherdel` would need to wait for the notification to trigger again, before the object is actually freed? I guess that could be surprising for someone watching memory usage after calling `dictwatcherdel`.

Comment (bfredl) [reply src/nvim/eval/typval.c]: I don't see the problem, is this not synchronous already? To me it looks like this free call will free the dict watcher in the _same_ event as the `callback_call` which caused it.

Comment (blueyed) [reply src/nvim/eval/typval.c]: Yes.

**Comment by erw7** [comment]
> Any ideas about the failiures still?

`w(QUEUE *)` exists within the `watcher(DictWatcher *w)`. Thus, if `tv_dict_watcher_free()` is called in `QUEUE_FOREACH`, `q` refers to the freed memory when evaluating `(q)! = (h)`. You will probably need to change it to:

```C
--- a/src/nvim/eval/typval.c
+++ b/src/nvim/eval/typval.c
@@ -1139,10 +1139,10 @@ bool tv_dict_watcher_remove(dict_T *const dict, const char *const key_pattern,
     return false;
   }

-  QUEUE_REMOVE(w);
   if (watcher->busy) {
     watcher->needs_free = true;
   } else {
+    QUEUE_REMOVE(w);
     tv_dict_watcher_free(watcher);
   }
   return true;
@@ -1205,8 +1205,9 @@ void tv_dict_watcher_notify(dict_T *const dict, const char *const key,

   typval_T rettv;

-  QUEUE *w;
-  QUEUE_FOREACH(w, &dict->watchers) {
+  QUEUE *w = QUEUE_HEAD(&dict->watchers);
+  while(1) {
+    QUEUE *next = w->next;
     DictWatcher *watcher = tv_dict_watcher_node_data(w);
     if (!watcher->busy && tv_dict_watcher_matches(watcher, key)) {
       rettv = TV_INITIAL_VALUE;
@@ -1214,10 +1215,15 @@ void tv_dict_watcher_notify(dict_T *const dict, const char *const key,
       callback_call(&watcher->callback, 3, argv, &rettv);
       watcher->busy = false;
       if (watcher->needs_free) {
+        QUEUE_REMOVE(w);
         tv_dict_watcher_free(watcher);
       }
       tv_clear(&rettv);
     }
+    if (next == &dict->watchers) {
+      break;
+    }
+    w = next;
   }

   for (size_t i = 1; i < ARRAY_SIZE(argv); i++) {
```

I think there is no problem in calling `QUEUE_REMOVE (w)` with `tv_dict_watcher_remove()`.

**Comment by blueyed** [comment]
@erw7 Thanks!

Should we rewrite the `QUEUE_FOREACH` macro like that?

**Comment by erw7** [comment]
My patch can be simpler.

```C
--- a/src/nvim/eval/typval.c
+++ b/src/nvim/eval/typval.c
@@ -1205,8 +1205,9 @@ void tv_dict_watcher_notify(dict_T *const dict, const char *const key,

   typval_T rettv;

-  QUEUE *w;
-  QUEUE_FOREACH(w, &dict->watchers) {
+  QUEUE *w = QUEUE_HEAD(&dict->watchers);
+  while(w != &dict->watchers) {
+    QUEUE *next = w->next;
     DictWatcher *watcher = tv_dict_watcher_node_data(w);
     if (!watcher->busy && tv_dict_watcher_matches(watcher, key)) {
       rettv = TV_INITIAL_VALUE;
@@ -1218,6 +1219,7 @@ void tv_dict_watcher_notify(dict_T *const dict, const char *const key,
       }
       tv_clear(&rettv);
     }
+    w = next;
   }

   for (size_t i = 1; i < ARRAY_SIZE(argv); i++) {
```

**Comment by erw7** [comment]
> Should we rewrite the QUEUE_FOREACH macro like that? Not possible due to the inner while loop though, is it?

Macros can be defined as follows:

```C
--- a/src/nvim/lib/queue.h
+++ b/src/nvim/lib/queue.h
@@ -39,6 +39,15 @@ typedef struct _queue {
   for (  /* NOLINT(readability/braces) */ \
       (q) = (h)->next; (q) != (h); (q) = (q)->next)

+#define QUEUE_WHILE(q, h, code) \
+  (q) = (h)->next; \
+  while((q) != (h)) { \
+    QUEUE *next = q->next; \
+    code \
+    (q) = next; \
+  }
+
+
 // ffi.cdef is unable to swallow `bool` in place of `int` here.
 static inline int QUEUE_EMPTY(const QUEUE *const q)
   FUNC_ATTR_ALWAYS_INLINE FUNC_ATTR_PURE FUNC_ATTR_WARN_UNUSED_RESULT
```

Other than this, there are two places where `while` is used to release memory in the loop. If we define a `QUEUE_WHILE` macro like the one above, I feel like it can be replaced.

https://github.com/neovim/neovim/blob/c5c06665ed302e774dc1a01bed949815b4b7c8c3/src/nvim/event/multiqueue.c#L122-L131

https://github.com/neovim/neovim/blob/c5c06665ed302e774dc1a01bed949815b4b7c8c3/src/nvim/os/pty_process_win.c#L297-L309

**Comment by blueyed** [comment]
Awesome.
Couldn't this replace `QUEUE_FOREACH` then?  (to make it safe by default)
I assume there is no more overhead in practice, is there?

**Comment by erw7** [comment]
> Couldn't this replace QUEUE_FOREACH then?

I have only confirmed that the compilation passes, but we can replace the `QUEUE_FOREACH` macro as follows:

https://github.com/blueyed/neovim/compare/fix-dictwatcherdel-crash...erw7:fix-dictwatcherdel-crash

> I assume there is no more overhead in practice, is there?

There should probably not be enough overhead to be a problem.

**Comment by blueyed** [comment]
@erw7 
I think your changes make sense.  Do you want to create a PR for it, or should I just pull them in here?

**Comment by erw7** [comment]
> Update: merged your branch here.

@blueyed thanks!

**Review by justinmk — APPROVED** [review]

**Comment by blueyed** [comment]
Reverted needs_free, not necessary anymore.

btw: I've noticed an issue with the syntax highlighting for the following, which marks the outer "{" as `cErrInParen` (cErrInParen cBadBlock cParen cBlock), and still another when using `:let g:c_no_curly_error = 1` (https://github.com/vim/vim/issues/3551).

**Comment by justinmk** [comment]
> btw: I've noticed an issue with the syntax highlighting for the following, which marks the outer "{" as `cErrInParen` (cErrInParen cBadBlock cParen cBlock), and still another when using `:let g:c_no_curly_error = 1` ([vim/vim#3551](https://github.com/vim/vim/issues/3551)).

Yep, known issue. Not really worth our time to fix it, if treesitter works out better.

File: src/nvim/eval/typval.c [hunk 3]
```diff
@@ -1201,16 +1201,15 @@ void tv_dict_watcher_notify(dict_T *const dict, const char *const key,
...
   QUEUE *w;
-  QUEUE_FOREACH(w, &dict->watchers) {
+  QUEUE_FOREACH(w, &dict->watchers, {
     DictWatcher *watcher = tv_dict_watcher_node_data(w);
     if (!watcher->busy && tv_dict_watcher_matches(watcher, key)) {
       rettv = TV_INITIAL_VALUE;
       watcher->busy = true;
       callback_call(&watcher->callback, 3, argv, &rettv);
       watcher->busy = false;
-      tv_clear(&rettv);
```
Comment (erw7) [inline src/nvim/eval/typval.c]: Oops! `tv_clear(&rettv)`; seems to be leaking in my commit https://github.com/neovim/neovim/pull/11189/commits/54ce37f5698ce6471adf9b373d1dd146c65b29dc.

**Comment by erw7** [comment]
> Reverted needs_free, not necessary anymore.

We still need `needs_free`. If you release the node in the callback, `watcher` does not point to valid memory at `watcher->busy = false;`.

https://github.com/neovim/neovim/blob/df31200a6c8a87df092c4895e02fa65486d819ac/src/nvim/eval/typval.c#L1209-L1210

File: src/nvim/eval/typval.c (hunk 3 above)
Comment (blueyed) [reply src/nvim/eval/typval.c]: Added in a4436ba - is that correct?

**Comment by blueyed** [comment]
> > Reverted needs_free, not necessary anymore.
> 
> We still need `needs_free`. If you release the node in the callback, `watcher` does not point to valid memory at `watcher->busy = false;`.
> 
> https://github.com/neovim/neovim/blob/df31200a6c8a87df092c4895e02fa65486d819ac/src/nvim/eval/typval.c#L1209-L1210

Ok, reverted.  I've only tested it using the test I've added.

**Comment by blueyed** [comment]
> Nice! Is this ready then? Would be good to squash such that @erw7 is author of the relevant commit.

Will do.

File: src/nvim/eval/typval.c (hunk 3 above)
Comment (erw7) [reply src/nvim/eval/typval.c]: Yes, that ’s fine.

**Comment by blueyed** [comment]
Squashed into two commits.

**Comment by blueyed** [comment]
There's still a problem here though (need to come up with a test): `tv_dict_watcher_remove` might free a non-current entry from the queue in https://github.com/blueyed/neovim/blob/4efb68605f61cc59a260be4e31a9900f3fddee13/src/nvim/eval/typval.c#L1146 still.
Pushed an attempt to address this: `tv_dict_watcher_remove` detects if the queue is busy altogether if any watcher is busy (might be better to set this on the dict maybe?), and then `tv_dict_watcher_notify` loops over the whole queue afterwards for removing/freeing.

**Comment by erw7** [comment]
@blueyed Please pull https://github.com/erw7/neovim/commit/e82006805a215c28c9519b258260c410b7a0b6d4. Because of my patch, when a callback(`W2`) was registered in the callback(`W1`) in the following cases, the callback(`W2`) was not called.

```vim
let g:d={}

function! W1(...) abort
  call dictwatcheradd(g:d, '*', function('W2'))
endfunction

function! W2(...) abort
  echomsg 'W2 was called'
endfunction

call dictwatcheradd(g:d, '*', function('W1'))
let g:d.foo = 123
```

**Comment by erw7** [comment]
Or since it is no longer necessary with https://github.com/neovim/neovim/pull/11189/commits/6e6af2779ff0d31b74a1c2684c4d798480abab92, it may be better to revert https://github.com/neovim/neovim/pull/11189/commits/4efb68605f61cc59a260be4e31a9900f3fddee13.

**Comment by blueyed** [comment]
> it may be better to revert [4efb686](https://github.com/neovim/neovim/commit/4efb68605f61cc59a260be4e31a9900f3fddee13).

I think it is a good change by itself, and would rather keep it.

I am not sure if a watcher added in the active callback should be called already?  It was not registered when the dict was changed after all.  (6e6af27 does not change that, i.e. your example still does echo nothing).

Should be added as a test when we decide how it should behave.

**Comment by justinmk** [comment]
is `any_needs_free` just a perf optimization? I would suggest removing that, because (1) watchers list is usually very small, and (2) it reduces "coverage" of that codepath.

**Comment by erw7** [comment]
> I am not sure if a watcher added in the active callback should be called already?

I have no use cases, but the behavior of this branch is confusing:

```vim
let g:d = {}
let g:a = []

function! W1(...) abort
  call dictwatcheradd(g:d, '*', function('W4'))
endfunction

function! W2(...) abort
endfunction

function! W3(...) abort
  call dictwatcheradd(g:d, '*', function('W5'))
endfunction

function! W4(...) abort
  call add(g:a, 'W4 was called')
endfunction

function! W5(...) abort
  call add(g:a, 'W5 was called')
endfunction

call dictwatcheradd(g:d, '*', function('W1'))
let g:d.foo = 123  " this branch: g:a == [], master: g:a == ['W4 was called']

unlet g:d
let g:d = {}
let g:a = []
call dictwatcheradd(g:d, '*', function('W1'))
call dictwatcheradd(g:d, '*', function('W2'))
let g:d.foo = 123  "this branch and master: g:a == ['W4 was called']

unlet g:d
let g:d = {}
let g:a = []
call dictwatcheradd(g:d, '*', function('W1'))
call dictwatcheradd(g:d, '*', function('W3'))
let g:d.foo = 123  "this branch and master: g:a == ['W4 was called', 'W5 was called']

unlet g:d
let g:d = {}
let g:a = []
call dictwatcheradd(g:d, '*', function('W2'))
call dictwatcheradd(g:d, '*', function('W3'))
let g:d.foo = 123  "this branch: g:a == [], master: g:a == ['W5 was called']
```

I think that it is better to make the same behavior as the master by modifying it as follows(https://github.com/erw7/neovim/commit/e82006805a215c28c9519b258260c410b7a0b6d4).

```diff
diff --git a/src/nvim/eval/typval.c b/src/nvim/eval/typval.c
index e90e4117a..36875bb9d 100644
--- a/src/nvim/eval/typval.c
+++ b/src/nvim/eval/typval.c
@@ -1211,7 +1211,7 @@ void tv_dict_watcher_notify(dict_T *const dict, const char *const key,

   bool any_needs_free = false;
   QUEUE *w;
-  QUEUE_FOREACH(w, &dict->watchers, {
+  for(w = (&dict->watchers)->next; w != &dict->watchers; w = w->next) {
     DictWatcher *watcher = tv_dict_watcher_node_data(w);
     if (!watcher->busy && tv_dict_watcher_matches(watcher, key)) {
       rettv = TV_INITIAL_VALUE;
@@ -1223,7 +1223,7 @@ void tv_dict_watcher_notify(dict_T *const dict, const char *const key,
         any_needs_free = true;
       }
     }
-  })
+  }
   if (any_needs_free) {
     QUEUE_FOREACH(w, &dict->watchers, {
       DictWatcher *watcher = tv_dict_watcher_node_data(w);
```

**Comment by erw7** [comment]
> Should be added as a test when we decide how it should behave.

I think that this branch can be used as it is, but it is better to create a test for the following cases to clarify the behavior.

```vim
let g:d = {}
let g:a = []

function! W1(...) abort
  call dictwatcherdel(g:d, '*', function('W2'))
endfunction

function! W2(...) abort
  call add(g:a, 'W2 was called')
endfunction

call dictwatcheradd(g:d, '*', function('W1'))
call dictwatcheradd(g:d, '*', function('W2'))
let g:d.foo = 123  " this branch: g:a == ['W2 was called'], master: g:a == []

unlet g:d
let g:d = {}
call dictwatcheradd(g:d, '*', function('W2'))
call dictwatcheradd(g:d, '*', function('W1'))
let g:d.foo = 123  " this branch and master: g:a == ['W2 was called']
```

**Comment by justinmk** [comment]
> I am not sure if a watcher added in the active callback should be called already? It was not registered when the dict was changed after all. ([6e6af27](https://github.com/neovim/neovim/commit/6e6af2779ff0d31b74a1c2684c4d798480abab92) does not change that, i.e. your example still does echo nothing).

Whatever results in the simplest implementation :)

**Comment by blueyed** [comment]
> is `any_needs_free` just a perf optimization? I would suggest removing that, because (1) watchers list is usually very small, and (2) it reduces "coverage" of that codepath.

Yes, for optimization.  I think it makes sense also with small lists in general, and we should not sacrifize anything for coverage purposes, but rather cover it then (assuming it only handles one branch there currently).

**Comment by justinmk** [comment]
> Actually both branches are covered:

I don't doubt it, but  removing it makes it "more covered". And the optimization is almost certainly negligible.

**Comment by blueyed** [comment]
> I don't doubt it, but removing it makes it "more covered".

No.  You still have the macro there (which is only partially covered for (to me) unknown reasons.
Actually having the extra code (that is covered) increases the percentage of coverage.
Also, the (first line of the) macro (which is only partially covered in other places) is covered there?!  https://codecov.io/gh/neovim/neovim/compare/ab860cb5f6c749b937e0986167702a36e45000ca...5a7d127721b7054ef97439d54d4ba648051f9f0f/diff#D5-1229...1234

I'd not pay too much attention to the details with regard to partial branch coverage with the macro itself, and would also still keep it if it would show up wrongly.  In this case however we have all branches covered and the (minor) optimization.

I'm still not sure about the `queue_is_busy` detection, and if it should have a flag on the queue (or dict?) instead.

**Comment by justinmk** [comment]
> I'm still not sure about the `queue_is_busy` detection, and if it should have a flag on the queue (or dict?) instead.

If this is asking about behavior my suggestion is still to just do whatever is easiest to implement.

**Comment by blueyed** [comment]
Yeah, then that's that (since it there already).
But it detects it by checking if any of the watchers is busy, which kind of feel like a hack to me (and only works for when it comes from there - have not really checked if that's the only code path).

Given that I think this is ready for now.

**Comment by blueyed** [comment]
Can merge it manually tomorrow, squashing it into two commits already then I think.

**Comment by teto** [comment]
@blueyed feel free to rebase/squash, would be nice to get it in 0.5

**Comment by janlazo** [comment]
Fails on Windows builds. https://github.com/neovim/neovim/actions/runs/703459630

**Comment by erw7** [comment]
> Fails on Windows builds. https://github.com/neovim/neovim/actions/runs/703459630

I think you can build it if you make the following changes.

```diff
diff --git a/src/nvim/os/pty_process_win.c b/src/nvim/os/pty_process_win.c
index 502b7ffa5..94444e4d2 100644
--- a/src/nvim/os/pty_process_win.c
+++ b/src/nvim/os/pty_process_win.c
@@ -505,11 +505,11 @@ static int build_env_block(dict_T *denv, wchar_t **env_block)
   *env_block = xmalloc(sizeof(**env_block) * env_block_len);
   wchar_t *pos = *env_block;

-  QUEUE_FOREACH(q, &env_q) {
+  QUEUE_FOREACH(q, &env_q, {
     EnvNode *env_node = QUEUE_DATA(q, EnvNode, node);
     memcpy(pos, env_node->str, env_node->len * sizeof(*pos));
     pos += env_node->len;
-  }
+  })

   *pos = L'\0';
```

**Comment by janlazo** [comment]
@erw7 It works. Thank you.

Closing in favor of https://github.com/neovim/neovim/pull/14259

**Pull request closed without merging**